File: 1-100/77-combinations.cpp
vector<vector<int>> combine(int n, int k) {
    vector<vector<int>> ans;
    vector<int> comb;
    backtracking(1, n, k, ans, comb);
    return ans;
}

void backtracking(int i, int n, int k, vector<vector<int>> &ans, vector<int> &comb) {
    if (comb.size() == k) {
        ans.push_back(comb);
        return;
    }
    for (int j = i; j <= n; ++j) {
        comb.push_back(j);
        backtracking(j+1, n, k, ans, comb);
        comb.pop_back();
    }
}

File: python version/77-combinations.py
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        comb = []
        ans = []

        def backtracking(start):
            if len(comb)+n-start+1<k:
                return
            if len(comb) == k:
                ans.append(comb[:])
                return
            for i in range(start, n+1):
                comb.append(i)
                backtracking(i+1)
                comb.pop()
        
        backtracking(1)
        return ans
            